### PR TITLE
fix: off by one slice bug

### PR DESCRIPTION
**Pull request by SweeXordious**
## Checklist

- [ ] New and updated code has appropriate documentation
- [ ] New and updated code has new and/or updated testing
- [ ] Required CI checks are passing
- [ ] Visual proof for any user facing features like CLI or documentation updates
- [ ] Linked issues closed with keywords

## Summary by CodeRabbit

- **New Features**
	- Enhanced data slicing functionality to handle invalid range inputs and out-of-bounds scenarios.
- **Bug Fixes**
	- Fixed an issue where data elements were incorrectly assigned during slicing.
- **Tests**
	- Added new tests to validate the improved slicing functionality and error handling.

### CODE REVIEW

File: src/lib/verifier/DAVerifier.sol [hunk 1]
```diff
@@ -236,9 +236,12 @@ library DAVerifier {
     /// @param _end The ending of the range (exclusive).
     /// @return _ the sliced data.
     function slice(bytes[] memory _data, uint256 _begin, uint256 _end) internal pure returns (bytes[] memory) {
+        if (_begin > _end) {
+            revert("Invalid range: _begin is greater than _end");
+        }
         bytes[] memory out = new bytes[](_end-_begin);
         for (uint256 i = _begin; i < _end; i++) {
-            out[i] = _data[i];
+            out[i - _begin] = _data[i];
```
Comment (evan-forbes) [inline src/lib/verifier/DAVerifier.sol:244]: :eyes:

**Review by evan-forbes — APPROVED** [review]

File: src/lib/verifier/DAVerifier.sol [hunk 2]
```diff
@@ -236,9 +236,12 @@ library DAVerifier {
     /// @param _end The ending of the range (exclusive).
     /// @return _ the sliced data.
     function slice(bytes[] memory _data, uint256 _begin, uint256 _end) internal pure returns (bytes[] memory) {
+        if (_begin > _end) {
+            revert("Invalid range: _begin is greater than _end");
+        }
```
Comment (adlerjohn) [inline src/lib/verifier/DAVerifier.sol:241]: Hmmm did we want library code to revert vs returning some sort of error?

Comment (SweeXordious) [reply src/lib/verifier/DAVerifier.sol]: [e2daa12](https://github.com/celestiaorg/blobstream-contracts/pull/254/commits/e2daa12b23501802cba0ad6c5147df479203d1af)

File: src/lib/verifier/DAVerifier.sol [hunk 3]
```diff
@@ -235,11 +240,21 @@ library DAVerifier {
...
+        if (_begin > _end) {
+            return (_data, ErrorCodes.InvalidRange);
+        }
+        if (_begin > _data.length || _end > _data.length) {
+            return (_data, ErrorCodes.OutOfBoundsRange);
+        }
         bytes[] memory out = new bytes[](_end-_begin);
         for (uint256 i = _begin; i < _end; i++) {
-            out[i] = _data[i];
+            out[i - _begin] = _data[i];
```
Comment (adlerjohn) [inline src/lib/verifier/DAVerifier.sol:256]: Can you add a test that hits this codepath and would have caught the bug?

Comment (SweeXordious) [reply src/lib/verifier/DAVerifier.sol:256]: [a042ac6](https://github.com/celestiaorg/blobstream-contracts/pull/254/commits/a042ac6b1124fa9880c779b8ff655a02164e83f5)

**Review by evan-forbes — APPROVED** [review]
lgtm but I'll defer to others for solidity things

**Review by adlerjohn — APPROVED** [review]

**Pull request merged**